### PR TITLE
Ruby: add model for Dir.glob and other Dir methods

### PR DESCRIPTION
**Pull request by erik-krogh**
CVE-2020-8161: Recognize the sink.

[Evaluation was very uneventful](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/PR-10888-0-ruby/reports).

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/frameworks/Files.qll [hunk 1]
```diff
@@ -182,3 +182,39 @@ module FileUtils {
     override DataFlow::Node getAPermissionNode() { result = permissionArg }
   }
 }
+
+/**
+ * Classes and predicates for modeling the core `Dir` module.
+ */
+module Dir {
+  /**
+   * A call methods on `Dir` that operates on a path as it's first argument, and produces file-names.
```
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/frameworks/Files.qll:191]:
```suggestion
   * A call to a method on `Dir` that operates on a path as its first argument, and produces file-names.
```

File: ruby/ql/lib/codeql/ruby/frameworks/Files.qll [hunk 2]
```diff
@@ -182,3 +182,39 @@ module FileUtils {
...
+      this =
+        API::getTopLevelMember("Dir")
+            .getAMethodCall(["glob", "[]", "children", "each_child", "entries", "foreach"])
+    }
+
+    override DataFlow::Node getAPathArgument() { result = super.getArgument(0) }
+  }
+
+  /**
+   * A call to a method in `Dir` which operates on a path as it's first argument, considered as a `FileSystemAccess`.
```
Comment (nickrolfe) [inline ruby/ql/lib/codeql/ruby/frameworks/Files.qll:205]:
```suggestion
   * A call to a method on `Dir` that operates on a path as its first argument, considered as a `FileSystemAccess`.
```

**Review by nickrolfe — COMMENTED** [review]
Looks sensible.

**Review by aibaars — COMMENTED** [review]
Looks good to me, one question though.

File: ruby/ql/lib/codeql/ruby/frameworks/Files.qll [hunk 3]
```diff
@@ -182,3 +182,39 @@ module FileUtils {
...
+
+/**
+ * Classes and predicates for modeling the core `Dir` module.
+ */
+module Dir {
+  /**
+   * A call to a method on `Dir` that operates on a path as its first argument, and produces file-names.
+   * Considered as a `FileNameSource` and a `FileSystemAccess`.
+   */
+  class DirGlob extends FileSystemAccess::Range, FileNameSource instanceof DataFlow::CallNode {
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply ruby/ql/lib/codeql/ruby/frameworks/Files.qll:194]: That was random.  
I like to use `instanceof` when appropriate, but I just didn't notice it in `DirPathAccess`.  

I've changed it, and done a merge with `main`.

**Review by aibaars — APPROVED** [review]

**Pull request merged**